src/codec.rs

```diff
@@ -13,3 +13,346 @@
+    //! }
+    //! #
+    //! # pub fn main() {}
+    //! ```
+    //!
+    //! The encoded frame will look like this:
+    //!
+    //! ```text
+    //! +---- len: u32 ----+---- data ----+
+    //! | \x00\x00\x00\x0b |  hello world |
+    //! +------------------+--------------+
+    //! ```
+    //!
+    //! # Decoding
+    //!
+    //! [`FramedRead`] adapts an [`AsyncRead`] into a `Stream` of [`BytesMut`],
+    //! such that each yielded [`BytesMut`] value contains the contents of an
+    //! entire frame. There are many configuration paramaters enabling
+    //! [`FrameRead`] to handle a wide range of protocols. Here are some
+    //! examples that will cover the various options at a high level.
+    //!
+    //! ## Example 1
+    //!
+    //! The following will parse a `u16` length field at offset 0, including the
+    //! frame head in the yielded `BytesMut`.
+    //!
+    //! ```
+    //! # use tokio_io::AsyncRead;
+    //! # use tokio_io::codec::length_delimited;
+    //! # fn bind_read<T: AsyncRead>(io: T) {
+    //! length_delimited::Builder::new()
+    //!     .length_field_offset(0) // default value
+    //!     .length_field_length(2)
+    //!     .length_adjustment(0)   // default value
+    //!     .num_skip(0) // Do not strip frame header
+    //!     .new_read(io);
+    //! # }
+    //! ```
+    //!
+    //! The following frame will be decoded as such:
+    //!
+    //! ```text
+    //!          INPUT                           DECODED
+    //! +-- len ---+--- Payload ---+     +-- len ---+--- Payload ---+
+    //! | \x00\x0B |  Hello world  | --> | \x00\x0B |  Hello world  |
+    //! +----------+---------------+     +----------+---------------+
+    //! ```
+    //!
+    //! The value of the length field is 11 (`\x0B`) which represents the length
+    //! of the payload, `hello world`. By default, [`FramedRead`] assumes that
+    //! the length field represents the number of bytes that **follows** the
+    //! length field. Thus, the entire frame has a length of 13: 2 bytes for the
+    //! frame head + 11 bytes for the payload.
+    //!
+    //! ## Example 2
+    //!
+    //! The following will parse a `u16` length field at offset 0, omitting the
+    //! frame head in the yielded `BytesMut`.
+    //!
+    //! ```
+    //! # use tokio_io::AsyncRead;
+    //! # use tokio_io::codec::length_delimited;
+    //! # fn bind_read<T: AsyncRead>(io: T) {
+    //! length_delimited::Builder::new()
+    //!     .length_field_offset(0) // default value
+    //!     .length_field_length(2)
+    //!     .length_adjustment(0)   // default value
+    //!     // `num_skip` is not needed, the default is to skip
+    //!     .new_read(io);
+    //! # }
+    //! ```
+    //!
+    //! The following frame will be decoded as such:
+    //!
+    //! ```text
+    //!          INPUT                        DECODED
+    //! +-- len ---+--- Payload ---+     +--- Payload ---+
+    //! | \x00\x0B |  Hello world  | --> |  Hello world  |
```

Comment:
Could this be swapped with the first example? This one at least seems like the first I'd run into, whereas the first is a bit more "advanced"